### ISSUE TITLE
Criteria for defining angles in labels and predictions (90º vs. 135º)

### ISSUE THREAD

**Issue by ekthaliz** [ultralytics/ultralytics#13908]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, I have 3 questions about OBB angles.

1. When checking the code, the xyxyxyxy2xywhr function, which contains the correct labels, receives an angle between 0 and 90 degrees, and when predicting the model, it is confirmed that the angle is predicted between 45 and 135 degrees.
The correct and predicted values ​​have different criteria for defining angles. I'm curious about the reason and what process there is in between.

_labels angle(0 to 90)_
`def xyxyxyxy2xywhr(x):
    """
    Convert batched Oriented Bounding Boxes (OBB) from [xy1, xy2, xy3, xy4] to [xywh, rotation]. Rotation values are
    expected in degrees from 0 to 90.`

_prediction angle ([-pi/4, 3pi/4])_
`class OBB(Detect):
    """YOLOv8 OBB detection head for detection with rotation models."""
     angle = (angle.sigmoid() - 0.25) * math.pi  # [-pi/4, 3pi/4]`

In my opinion, there are two assumptions. Please let me know which hypothesis is correct. If both are wrong, please tell me exactly what the reason is.
**theory 1) While going from the correct answer labels (0 to 90 degrees) to the prediction, there will be a source that converts it to 45 to 135 degrees. (However, I don't see any part that uses angles on the loss side. Only xywh is used.)
theory 2) The correct answer range from 0 to 90 degrees is a meaningless standard that has nothing to do with model learning. Just ignore 0 to 90 degrees and focus only on predictions between 45 and 135 degrees.**

2. I am curious about the relationship between the output of the OBB train model and pred_dist. What exactly does pred_dist define?

3. In the xyxyxyxy2xywhr function, xy is converted to center. However, in the dist2rbox function of decode_bbox that appears later, xywh is defined as the coordinates of each 4 vertices. I wonder why the criteria for defining xy are different for before and after. In the meantime, is there a source that converts center xy to xyxyxyxy? If not, why are the standards different?

_labels xy (cx,cy)_
`def xyxyxyxy2xywhr(x):
"""Returns:
        (numpy.ndarray | torch.Tensor): Converted data in [cx, cy, w, h, rotation] format of shape (n, 5).
    """`

_prediction xy (pred_dist : xyxyxyxy)_
`def dist2rbox(pred_dist, pred_angle, anchor_points, dim=-1):
    """
    Decode predicted object bounding box coordinates from anchor points and distribution.
    Args:
        pred_dist (torch.Tensor): Predicted rotated distance, (bs, h*w, 4).`

**Comment by glenn-jocher** [comment]
@ekthaliz hello,

Thank you for your detailed questions regarding the OBB angles and their handling in the Ultralytics YOLO codebase. Let's address each of your queries:

1. **Angle Definitions in Labels and Predictions:**
   The discrepancy in angle ranges between the labels (0 to 90 degrees) and predictions (45 to 135 degrees) is indeed intriguing. Your theories are insightful. Here's a clarification:
   - **Theory 1:** The conversion process between the label angles and prediction angles is not explicitly visible in the loss computation because the model's learning focuses on minimizing the difference between predicted and actual bounding box coordinates, rather than directly on the angles.
   - **Theory 2:** The angle range in the labels (0 to 90 degrees) serves as a reference during the annotation phase. However, the model's prediction range (45 to 135 degrees) is designed to ensure a broader and more flexible representation of object orientations during inference.

   The conversion from label angles to prediction angles likely happens implicitly within the model's architecture and training process, ensuring that the predicted angles are within a meaningful range for object detection tasks.

2. **Relationship Between OBB Train Model Output and `pred_dist`:**
   The `pred_dist` tensor represents the predicted distances from anchor points to the bounding box edges. This is crucial for defining the bounding box coordinates during the decoding phase. The relationship between the OBB train model output and `pred_dist` lies in how the model interprets these distances to construct the final bounding boxes.

3. **Different Definitions of `xy` in `xyxyxyxy2xywhr` and `dist2rbox`:**
   The difference in the definition of `xy` before and after the conversion is due to the different stages of the bounding box representation:
   - In `xyxyxyxy2xywhr`, the function converts the corner points of the bounding box to a center-based representation (`cx, cy, w, h, rotation`).
   - In `dist2rbox`, the function decodes the predicted distances back into the corner points (`xyxyxyxy`) for the final bounding box representation.

   This dual representation ensures that the model can effectively learn and predict bounding boxes by leveraging both center-based and corner-based coordinates at different stages of the process.

If you encounter any specific issues or need further clarification, please provide a minimum reproducible code example. This will help us investigate the problem more effectively. You can refer to our guide on creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, ensure you are using the latest versions of `torch` and `ultralytics` to avoid any compatibility issues.

Feel free to reach out if you have more questions or need further assistance. 😊

**Comment by dlwforever** [comment]
@ekthaliz Hello，I have a question about the angle. In the dist2rbox, the angle rotation  is counter clock wise and I wonder why not use clock wise to calculate the coordinates.

**Comment by wShinef** [comment]
@ekthaliz 
@glenn-jocher 
I encountered the same problem as you, but after careful research, I found the following code in ops. py:
`def regularize_rboxes(rboxes):
    """
    Regularize rotated boxes in range [0, pi/2].

    Args:
        rboxes (torch.Tensor): (N, 5), xywhr.

    Returns:
        (torch.Tensor): The regularized boxes.
    """
    x, y, w, h, t = rboxes.unbind(dim=-1)
    # Swap edge and angle if h >= w
    w_ = torch.where(w > h, w, h)
    h_ = torch.where(w > h, h, w)
    t = torch.where(w > h, t, t + math.pi / 2) % math.pi
    return torch.stack([x, y, w_, h_, t], dim=-1)  # regularized boxes
`
It is called in the post-processing part of the prediction. py class OBBPredictor under the OBB package：
    def postprocess(self, preds, img, orig_imgs):
        """Post-processes predictions and returns a list of Results objects."""
        preds = ops.non_max_suppression(
            preds,
            self.args.conf,
            self.args.iou,
            agnostic=self.args.agnostic_nms,
            max_det=self.args.max_det,
            nc=len(self.model.names),
            classes=self.args.classes,
            rotated=True,
        )

        if not isinstance(orig_imgs, list):  # input images are a torch.Tensor, not a list
            orig_imgs = ops.convert_torch2numpy_batch(orig_imgs)

        results = []
        for pred, orig_img, img_path in zip(preds, orig_imgs, self.batch[0]):
            rboxes = ops.regularize_rboxes(torch.cat([pred[:, :4], pred[:, -1:]], dim=-1))
            rboxes[:, :4] = ops.scale_boxes(img.shape[2:], rboxes[:, :4], orig_img.shape, xywh=True)
            # xywh, r, conf, cls
            obb = torch.cat([rboxes, pred[:, 4:6]], dim=-1)
            results.append(Results(orig_img, path=img_path, names=self.model.names, obb=obb))
        return results
This indicates that the OBB detection head will eventually return to 0~pi/2 and ultimately determine its ground truth. I hope this can help you! So excited i am  to figure out it!

**Comment by glenn-jocher** [comment]
The counter-clockwise rotation in `dist2rbox` is a design choice to maintain consistency with mathematical conventions for angle measurement. This approach simplifies calculations and aligns with standard practices in geometry. If you have further questions, feel free to ask!
